tests/websocket.c

```diff
@@ -8,319 +8,537 @@
-	ret = write(fd, message, sizeof(message)-1);
-	if(ret != sizeof(message)-1) {
-		fprintf(stderr, "write on %d failed: %s\n", fd, strerror(errno));
-		close(fd);
-	}
-
-	wt->msg_sent += 2;
-	if(wt->msg_sent < wt->msg_target) {
-		event_set(&wt->ev_w, fd, EV_WRITE, websocket_write, wt);
-		event_base_set(wt->base, &wt->ev_w);
-		ret = event_add(&wt->ev_w, NULL);
-	}
+websocket_can_write(int fd, short event, void *ptr) {
+    int ret;
+    struct worker_thread *wt = ptr;
+    wt->debug("%s (wt=%p, fd=%d)\n", __func__, wt, fd);
+
+    if(event != EV_WRITE) {
+        return;
+    }
+    switch (wt->state)
+    {
+    case WS_INITIAL: { /* still sending initial HTTP request */
+        ret = evbuffer_write(wt->wbuffer, fd);
+        wt->debug("evbuffer_write returned %d\n", ret);
+        wt->debug("evbuffer_get_length returned %d\n", evbuffer_get_length(wt->wbuffer));
+        if (evbuffer_get_length(wt->wbuffer) != 0) { /* not all written */
+            wait_for_possible_write(wt);
+            return;
+        }
+        /* otherwise, we've sent the full request, time to read the response */
+        wt->state = WS_SENT_HANDSHAKE;
+        wt->debug("state=WS_SENT_HANDSHAKE\n");
+        wait_for_possible_read(wt);
+        return;
+    }
+    case WS_RECEIVED_HANDSHAKE: { /* ready to send a frame */
+        wt->debug("About to send data for WS frame, %lu in buffer\n", evbuffer_get_length(wt->wbuffer));
+        evbuffer_write(wt->wbuffer, fd);
+        size_t write_remains = evbuffer_get_length(wt->wbuffer);
+        wt->debug("Sent data for WS frame, still %lu left to write\n", write_remains);
+        if (write_remains == 0) { /* ready to read response */
+            wt->state = WS_SENT_FRAME;
+            wt->msg_sent++;
+            wait_for_possible_read(wt);
+        } else { /* not finished writing */
+            wait_for_possible_write(wt);
+        }
+        return;
+    }
+    default:
+        break;
+    }
 }
 
 static void
-websocket_read(int fd, short event, void *ptr) {
-	char packet[2048], *pos;
-	int ret, success = 1;
-
-	struct worker_thread *wt = ptr;
-
-	if(event != EV_READ) {
-		return;
-	}
-
-	/* read message */
-	ret = read(fd, packet, sizeof(packet));
-	pos = packet;
-	if(ret > 0) {
-		char *data, *last;
-		int sz, msg_sz;
-
-		if(wt->got_header == 0) { /* first response */
-			char *frame_start = strstr(packet, "MH"); /* end of the handshake */
-			if(frame_start == NULL) {
-				return; /* not yet */
-			} else { /* start monitoring possible writes */
-				printf("start monitoring possible writes\n");
-				evbuffer_add(wt->buffer, frame_start + 2, ret - (frame_start + 2 - packet));
-
-				wt->got_header = 1;
-				event_set(&wt->ev_w, fd, EV_WRITE,
-						websocket_write, wt);
-				event_base_set(wt->base, &wt->ev_w);
-				ret = event_add(&wt->ev_w, NULL);
-			}
-		} else {
-			/* we've had the header already, now bufffer data. */
-			evbuffer_add(wt->buffer, packet, ret);
```

Comment:
so this _was_ already using libevent buffers?

Reply:
part of it, yes.